client/src/components/Modal/tests/Modal-test.js

```diff
@@ -0,0 +1,89 @@
+/* global jest, test, describe, it, expect */
+
+import React from 'react';
+import { fireEvent, render } from '@testing-library/react';
+import Modal from '../Modal';
+
+let props;
+
+/** Component that printouts all of its props */
+const PrintProps = (componentName) => ({ children, ...propsToPrint }) => (
+  <div>
+    <div>{componentName}</div>
+    <ul>
+      {Object
+        .entries(propsToPrint)
+        .map(([key, value]) => (<li key={key}>{key}: {value.toString()}</li>))
+      }
+    </ul>
+    {children}
+  </div>
+);
+
+beforeEach(() => {
+  props = {
+    isOpen: true,
+    className: 'my-modal',
+    modalClassName: 'my-modal-dialog',
+    size: 'sm',
+    onClosed: jest.fn(),
+    title: 'Hello World!',
+    showCloseButton: true,
+  };
+});
+
+test('Modal renders', () => {
+  const root = render(
+    <Modal {...props}>My Content</Modal>
+  );
+
+  const modal = root.getByRole('dialog');
+  expect(modal.style.getPropertyValue('display')).toBe('block');
+  expect(modal.classList).toContain('my-modal-dialog');
+  expect(modal.classList).toContain('modal');
+
+  const document = root.getByRole('document');
+  expect(document.classList).toContain('modal-sm');
+
+  const header = root.getByRole('heading');
+  expect(header.textContent).toContain('Hello World!');
+
+  const body = root.getByText('My Content');
+  expect(body).toBeTruthy();
+
+  const closeButton = root.getByLabelText('Close');
+  expect(closeButton).toBeTruthy();
+});
+
+test('Closing the Modal', () => {
+  const root = render(
+    <Modal {...props}>My Content</Modal>
+  );
+
+  const closeButton = root.getByLabelText('Close');
+  fireEvent.click(closeButton);
+
+  expect(props.onClosed).toBeCalled();
+});
+
+test('Modal is not shown', () => {
+  const root = render(
+    <Modal {...props} isOpen={false}>My Content</Modal>
+  );
+  expect(root.container.textContent).toBeFalsy();
```

Comment:
```suggestion
  expect(root.container.querySelectorAll('.some-class')).toHaveLength(0);
```
This is how we do this sort of thing elsewhere

Reply:
For some reason I'm not quite sure about, `root.container.querySelectorAll` never returns anything for any of my Modal tests.

While trying to debug why it wasn't working, I came across [this official guidance](https://testing-library.com/docs/queries/about/#priority) that strongly to avoid CSS selectors to find element from the virtual dom.

> your test should resemble how users interact with your code (component, page, etc.) as much as possible
> ...
> Queries that reflect the experience of visual/mouse users as well as those that use assistive technology.

Reply:
:-) That guidance also means you should be using things like `await screen.findByRole()` to sort out all the async fun present in javascript.  You're more than welcome to use that if you want, though container.querySelector() has been used throughout our codebase because it's far easier to work with.

The fact that container.querySelector() isn't working is kinda concerning, probably means that your old test wasn't actually testing what you thought it was

Testing negatives is hard. You should be able to use something like this to wait for everything to load:

```
import { render, act } from '@testing-library/react';
// ...
  await act(async () => {
    await new Promise((resolve) => setTimeout(resolve, 100));
  });
```

Reply:
When I run `root.debug()`, it prints out exactly the HTML I would expect.

```html
    <body
      class="modal-open"
    >
      <div />
      <div
        style="position: relative; z-index: 1050; display: block;"
        tabindex="-1"
      >
        <div
          class=""
        >
          <div
            class="modal my-modal-dialog fade"
            role="dialog"
            style="display: block;"
            tabindex="-1"
          >
            <div
              class="modal-dialog my-modal modal-sm"
              role="document"
            >
              <div
                class="modal-content"
              >
                <div
                  class="modal-header"
                >
                  <h5
                    class="modal-title"
                  >
                    Hello World!
                  </h5>
                  <button
                    aria-label="Close"
                    class="close"
                    type="button"
                  >
                    <span
                      aria-hidden="true"
                    >
                      ×
                    </span>
                  </button>
                </div>
                My Content
              </div>
            </div>
          </div>
          <div
            class="modal-backdrop fade"
          />
        </div>
      </div>
    </body>
```

I'm really not sure why `root.container.querySelector()` doesn't seem to work for this test. I'm really not sure what I'm doing different on this test compare to all the other ones.

Reply:
Your comment about testing negative was prescient. The test was passing whether `isOpen` was false or true.

I've update the test to use screen instead and it's failing when isOpen is true has you would expect.
```js
const modalBody = screen.queryByText('My Content');
expect(modalBody).toBeNull();
```